### PR TITLE
Ports bay's supermatter monitor modular program

### PR DESCRIPTION
**Pull request by Shadowlight213**
Holy shit, actually new content for modular computers instead of refactors!!! :tada: 

:cl:
add: There is now a new monitor program that engineers can use to monitor the supermatter status
/:cl:
Screenshot of the UI (the double colon thing was fixed. this one is a bit old)

Screen overlays! Laptops use the Eris static version for now. The others will change depending on supermatter status.

### CODE REVIEW

**Comment by Bawhoppen** [comment]
is this able to dynamically monitor any supermatter or just the roundstart engineering ones?

File: tgui/src/interfaces/ntos_supermatter_monitor.ract [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	  integState(charge) {
+	    let max = 100
+	    if(charge == max) return 'good'
+		else if(charge > (max/2)) return 'average'
+		else return 'bad'
+	    }
+      }
+    }
+
+  component.exports = {
```
Comment (RemieRichards) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]: you don't need a second set of ```component.exports``` and ```data```, just put ```BigState(val, aver, max)``` below ```integState```

Like I did here: https://github.com/tgstation/tgstation/blob/eace4eec148e8a366ee2fe6673f852b59a2e77b1/tgui/src/interfaces/mech_bay_power_console.ract#L2

**Comment by Shadowlight213** [comment]
Any supermatter. The program starts you on a list of all supermatters detected. It will show any that are on the station or mining level, or on the same level as the user. You can then select one to view in more detail.

File: tgui/src/interfaces/ntos_supermatter_monitor.ract (hunk 1 above)
Comment (Shadowlight213) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]: I did that at first, but it gave me compile errors.

**Comment by Bawhoppen** [comment]
It'd be better if you had to stick a monitoring chip on or something probably instead of just showing all

**Comment by Shadowlight213** [comment]
All supermatters have a monitoring chip on them though. The radio. IC wise, this just interfaces with those existing sensors.

**Comment by Bawhoppen** [comment]
I guess

File: tgui/src/interfaces/ntos_supermatter_monitor.ract (hunk 1 above)
Comment (Shadowlight213) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]: @RemieRichards for future reference, is there a way to do an else if, or switch statement in tgui?

Comment (RemieRichards) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]: in the template or in the JS at the top of templates?

Comment (Shadowlight213) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]: In the template

Comment (RemieRichards) [inline tgui/src/interfaces/ntos_supermatter_monitor.ract]:
```
{{#if data.example}}
{{else}}
    {{#if data.example_two}}
    {{/if}}
{{/if}}
```
You have to remember that ```else if``` is really just sugar for ```if {} else { if { } }```

**Comment by kevinz000** [comment]
>doing network refactor
>see this
>SCREAM

**Comment by meeeeme** [comment]
neat

**Comment by Anonmare** [comment]
Is MeV/cm^3 charge level?

**Comment by Shadowlight213** [comment]
yes

**Comment by Shadowlight213** [comment]
Ok, so I was able to codersprite the overlays for the program onto the tablet, but there is no way I'm managing to get them to fit onto the Eris laptop sprites. The laptop screens are just too small.

**Comment by Shadowlight213** [comment]
Kor won't let me use the bay laptops because I said mean words to oranges, and I cant fit the overlays on the eris ones.

**Comment by KorPhaeron** [comment]
????

**Comment by KorPhaeron** [comment]
Just use the new sprites you meme

**Comment by Shadowlight213** [comment]
They don't fit. The overlay is too big senpai. It's not going to fit.

**Comment by KorPhaeron** [comment]
Then use the bay laptops nobody cares what the laptop sprites are I promise I thought asking permission about that was a joke

**Comment by optimumtact** [comment]
maybe try the tip of the overlay?

**Comment by optimumtact** [comment]
just the tip

**Comment by lzimann** [comment]
@Shadowlight213

**Comment by Shadowlight213** [comment]
Ok, I have the overlay be overridden if a laptop is running the program, as I'd need more time to figure out how to port the bay overlays there.

**Comment by Shadowlight213** [comment]
Whoops.

**Comment by Shadowlight213** [comment]
This is basically done.

File: code/modules/modular_computers/computers/item/tablet_presets.dm [hunk 2]
```diff
@@ -19,6 +22,12 @@
 	install_component(new /obj/item/weapon/computer_hardware/network_card)
 	install_component(new /obj/item/weapon/computer_hardware/card_slot)
 	install_component(new /obj/item/weapon/computer_hardware/printer/mini)
+	install_programs()
+
+/obj/item/device/modular_computer/tablet/preset/advanced/engi/install_programs()
+	var/obj/item/weapon/computer_hardware/hard_drive/hard_drive = all_components[MC_HDD]
+	hard_drive.store_file(new/datum/computer_file/program/alarm_monitor())
+	hard_drive.store_file(new/datum/computer_file/program/supermatter_monitor())
```
Comment (Cyberboss) [inline code/modules/modular_computers/computers/item/tablet_presets.dm]: Can't say I'm a fan of pre-installing programs, tbh

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 3]
```diff
@@ -0,0 +1,136 @@
+/datum/computer_file/program/supermatter_monitor
+	filename = "smmonitor"
+	filedesc = "Supermatter Monitoring"
+	ui_header = "smmon_0.gif"
+	program_icon_state = "smmon_0"
+	extended_desc = "This program connects to specially calibrated supermatter sensors to provide information on the status of supermatter-based engines."
+	requires_ntnet = 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	transfer_access = GLOB.access_engine
+	network_destination = "supermatter monitoring system"
+	size = 5
+	tgui_id = "ntos_supermatter_monitor"
+	ui_x = 600
+	ui_y = 400
+	var/last_status = 0
+	var/list/supermatters
+	var/obj/machinery/power/supermatter_shard/active = null		// Currently selected supermatter crystal.
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: Remove ` = null`

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	var/list/data = get_header_data()
+
+	if(istype(active))
+		var/turf/T = get_turf(active)
+		if(!T)
+			active = null
+			refresh()
+			return
+		var/datum/gas_mixture/air = T.return_air()
+		if(!istype(air))
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: just !air

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			active = null
+			return
+
+		data["active"] = 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+		data["SM_integrity"] = active.get_integrity()
+		data["SM_power"] = active.power
+		data["SM_ambienttemp"] = air.temperature
+		data["SM_ambientpressure"] = air.return_pressure()
+		//data["SM_EPR"] = round((air.total_moles / air.group_multiplier) / 23.1, 0.01)
+		var/list/gasdata = list()
+		var/list/relevantgas = list("o2","co2","n2","plasma","n2o","freon")
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: why exclude anything? this is basically all gases

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		for(var/obj/machinery/power/supermatter_shard/S in supermatters)
+			var/area/A = get_area(S)
+			if(A)
+				SMS.Add(list(list(
+				"area_name" = A.name,
+				"integrity" = S.get_integrity(),
+				"uid" = S.uid
+				)))
+
+		data["active"] = 0
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: FALSE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		data["supermatters"] = SMS
+
+	return data
+
+/datum/computer_file/program/supermatter_monitor/ui_act(action, params)
+	if(..())
+		return 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+	switch(action)
+		if("PRG_clear")
+			active = null
+			return 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+		if("PRG_refresh")
+			refresh()
+			return 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/modular_computers/file_system/programs/sm_monitor.dm [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+		if("PRG_set")
+			var/newuid = text2num(params["set"])
+			for(var/obj/machinery/power/supermatter_shard/S in supermatters)
+				if(S.uid == newuid)
+					active = S
+			return 1
```
Comment (Cyberboss) [inline code/modules/modular_computers/file_system/programs/sm_monitor.dm]: TRUE

File: code/modules/power/supermatter/supermatter.dm [hunk 13]
```diff
@@ -165,6 +165,41 @@
...
+
+/obj/machinery/power/supermatter_shard/proc/get_status()
+	var/turf/T = get_turf(src)
+	if(!T)
+		return SUPERMATTER_ERROR
+	var/datum/gas_mixture/air = T.return_air()
+	if(!air)
+		return SUPERMATTER_ERROR
+
+	if(get_integrity() < 5)
```
Comment (Cyberboss) [inline code/modules/power/supermatter/supermatter.dm]: There should be defines for these levels

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/power/supermatter/supermatter.dm (hunk 13 above)
Comment (Shadowlight213) [inline code/modules/power/supermatter/supermatter.dm]: You sure? Its a percentage of the sm integrity remaining.

File: code/modules/modular_computers/computers/item/tablet_presets.dm (hunk 2 above)
Comment (Shadowlight213) [inline code/modules/modular_computers/computers/item/tablet_presets.dm]: I can remove it if you want.

Comment (Cyberboss) [inline code/modules/modular_computers/computers/item/tablet_presets.dm]: Please do, I believe tablets are eventually supposed to be a: build your own pda sort of deal

File: code/modules/power/supermatter/supermatter.dm (hunk 13 above)
Comment (Cyberboss) [inline code/modules/power/supermatter/supermatter.dm]: You're right, Can you make them defines please

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**